docs(nx-cloud): improve info on setting access tokens in various CI providers (#31261)
nrwl/nx: docs/nx-cloud/recipes/access-tokens.md

### Read Only Access

The `read-only` access tokens will only read from the remote cache. New task results will be stored in the remote cache _only_ for that specific branch in a CI context, otherwise they will not be stored in the remote cache.  
Cached results can be downloaded and replayed for other machines or CI pipelines to use. This option provides the benefit of remote cache hits while restricting machines without proper permissions from adding entries into the shared primary remote cache.

### Read & Write Access

The `read-write` access tokens allows task results to be stored in the remote cache for other machines or CI pipelines to download and replay. This access level should only be used for trusted environments, such as protected branches within your CI Pipeline.

## Setting CI Access Tokens

You can configure an access token in CI by setting the `NX_CLOUD_ACCESS_TOKEN` environment variable.

`NX_CLOUD_ACCESS_TOKEN` takes precedence over any authentication method in your `nx.json`.

> We recommend setting up a `read-write` and `read-only` in your CI based on protected vs unprotected branches. You can leverage your CI providers.

Reviewer: ```suggestion
We recommend setting up a `read-write` token for you protected branches in CI and a `read-only` token for unprotected branches and local development. You can leverage your CI provider's environment variables management to accomplish this.
```

### Azure DevOps

> Azure DevOps provides various [mechanisms to limit access to secrets](https://learn.microsoft.com/en-us/azure/devops/pipelines/security/secrets?view=azure-devops#limit-access-to-secret-variables). We'll be using _Variable groups_ in this process, but you can achieve the same result leveraging [Azure KeyVault](https://learn.microsoft.com/en-us/azure/key-vault/general/overview).

Reviewer: ```suggestion
Azure DevOps provides various [mechanisms to limit access to secrets](https://learn.microsoft.com/en-us/azure/devops/pipelines/security/secrets?view=azure-devops#limit-access-to-secret-variables). We'll be using _variable groups_ in this process, but you can achieve the same result leveraging [Azure Key Vault](https://learn.microsoft.com/en-us/azure/key-vault/general/overview).
```

Author: Azure UI shows Variable group as the feature and button name to do it. so leaving it as capitalized to make users expectation to the azure system. 

+1 on Key Vault

1. In your project, navigate to Pipelines > Library.
   ![Variable group settings page](/nx-cloud/recipes/ado-library-start.avif)

> 2. Create a new _Variable group_ called _protected_.

Reviewer: ```suggestion
2. Create a new _variable group_ called _protected_.
```

Author: Azure UI shows Variable group as the feature and button name to do it. so leaving it as capitalized to make users expectation to the azure system.

    ![unprotected variable group settings](/nx-cloud/recipes/ado-unprotected-branch.avif)
11. Now you should see 2 _Variable groups_ for _protected_ and _unprotected_ usage.
    ![completed variable group setup](/nx-cloud/recipes/ado-library-end.avif)
12. Update your pipeline to include the 2 variable groups, with conditional access for the _protected_ variable group.

Example usage:

```yaml {% fileName="azure-pipelines.yml" highlightLines=["2-4"] %}
variables:
  - group: unprotected
  - ${{ if eq(variables['Build.SourceBranchName'], 'main') }}:
      - group: protected
```

{% callout type="check" title="Can't someone change the variable group?" %}
Since we use the _Verify branch protection_ option, CI can only read the variable when running in a protected branch. If a developer tries to edit the pipeline to use the _protected_ variable group, the pipeline will error out since permissions require running in on a protected branch.

Take caution though, if you allow team members to have direct write access to a protected branch, then they could modify the pipeline to write to the nx cache without having a code review first.
{%/callout %}

### BitBucket Cloud

> BitBucket Cloud supports setting a environment variables per environment called _Deployment variables_. You can read the [official BitBucket Pipelines documentation](https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/#Deployment-variables) for more details.

Reviewer: ```suggestion
BitBucket Cloud supports setting environment variables per environment called _Deployment variables_. You can read the [official BitBucket Pipelines documentation](https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/#Deployment-variables) for more details.
```

   ![restrict context to protected branches](/nx-cloud/recipes/circle-expression-restriction.avif)
3. Click on _Add Environment Variable_ and add the `NX_CLOUD_ACCESS_TOKEN` environment variable with the `read-write` token from Nx Cloud.
4. Back on the organization home page, navigate to your projects, then view the pipeline settings.
5. Navigate to _Environment Variables_ and click _Add Environment Variable_ and add the `NX_CLOUD_ACCESS_TOKEN` environment variable with the `read-only` token from Nx Cloud.
   ![add read-only nx cloud access token to circleci](/nx-cloud/recipes/circle-new-env-var-token.avif)
6. Update your pipeline to include steps where you want to write to the nx cache with the correct contexts.

Example usage:

```yaml {% fileName=".circleci/config.yml" highlightLines=["10-19"] %}
jobs:
  run-tests-protected:
    - ...
  run-tests-prs:
    - ...

workflows:
  my-workflow:
    jobs:
      - run-tests-protected:
          context:
            - protected-branches
          filters:
            branches:
              only: main
      - run-tests-prs:
          filters:
            branches:
              ignore: main
```

### GitHub Actions

> GitHub allows specifying specific secrets for each environment, where an environment can be on a specific branch.

Reviewer: ```suggestion
GitHub allows specifying different secrets for each environment, where an environment can be on a specific branch.
```

### Read Only Access

The `read-only` access tokens can only read from the global remote cache. Task results produced with this type of access token will be stored in an isolated remote cache accessible _only_ by that specific branch in a CI context, and cannot influence the global shared cache.  
The isolated remote cache produced with a `read-only` token is accessible to all machines or agents in the same CI execution, enabling cache sharing during distributed task execution.

### Read & Write Access

The `read-write` access tokens allow task results to be stored in the remote cache for other machines or CI pipelines to download and replay. This access level should only be used for trusted environments such as protected branches within your CI Pipeline.

## Setting CI Access Tokens

You can configure an access token in CI by setting the `NX_CLOUD_ACCESS_TOKEN` environment variable.

> The `NX_CLOUD_ACCESS_TOKEN` takes precedence over any authentication method in your `nx.json`.

Reviewer 2: Should we be clear that this is an an env variable and provide an example of how to use it?

Author: all the changes are showing how to use per CI provider now instead of a blanket github action one.

    steps: ...
```

### GitLab

> GitLab allows creating variables scoped to specific variables. You can read the [Official GitLab documentation](https://docs.gitlab.com/ci/environments/#limit-the-environment-scope-of-a-cicd-variable) for more details.

Reviewer: ```suggestion
GitLab allows creating variables scoped to specific environments. You can read the [Official GitLab documentation](https://docs.gitlab.com/ci/environments/#limit-the-environment-scope-of-a-cicd-variable) for more details.
```

   ![add read-write nx cloud access token to gitlab](/nx-cloud/recipes/gitlab-variable-settings-readwrite.avif)
7. Now you should see 2 secrets where 1 is a part of the protected & tagged to the environment and the other is not.
   ![GitLab project variable configuration screen](/nx-cloud/recipes/gitlab-variable-setting.avif)
8. Update your pipeline to include steps where you want to write to the nx cache with the correct contexts.

Example usage:

```yaml {% fileName=".gitlab-ci.yml" highlightLines=["2-3"] %}
<job-name>:
  environment:
    name: <environment-name>
```

{% callout type="check" title="Can't someone change the step environment?" %}
Since we use the _Protected variable_ flag, CI can only read the variable when running in a protected branch. If a developer tries to edit the steps to run a PR in the environment with the `read-write` token will, then the token will not be populated in CI since their branch is not marked as protected.

Take caution though, if you allow team members to have direct write access to a protected branch, then they could modify the steps to write to the nx cache without having a code review first.
{%/callout %}

### Jenkins

> Jenkins configuration can be quite extensive make each Jenkins instance potentially different from each other. Because of this we can only provide a minimal viable approach, but there can be a multiple way to provide scoped access tokens to your pipelines. The goal is to create two areas within Jenkins, where one is the _protected_ and the other is the _unprotected_. These specifically map to how you deem your branches should have read/write vs read permissions. We recommend making branches that developers cannot directly push to and require a code review to merge to, as the _protected_ branches, and the rest being _unprotected_.

Reviewer: ```suggestion
Jenkins configuration can be quite extensive making each Jenkins instance unique. Because of this we can only provide a minimal viable approach, but there can be multiple ways to provide scoped access tokens to your pipelines. The goal is to create two areas within Jenkins, where one is the _protected_ and the other is the _unprotected_. These specifically map to how you deem your branches should have read/write vs read permissions. We recommend making branches that developers cannot directly push to and require a code review to merge to, as the _protected_ branches, and the rest being _unprotected_.
```
